lib/elixir/lib/kernel/typespec.ex

```diff
@@ -478,11 +478,46 @@ defmodule Kernel.Typespec do
   ## Macro callbacks
 
   @doc false
+  def defspec(:callback, expr, caller) do
+    {name, arity} = spec_to_signature(expr)
+
+    name_str = Atom.to_string(name)
+
+    {docs_kind, docs_name, docs_arity} =
+      if match?("MACRO-" <> _, name_str) do
+        "MACRO-" <> macro_name = name_str
+        {:defmacro, String.to_atom(macro_name), arity - 1}
+      else
+        {:def, name, arity}
+      end
+
+    Module.store_typespec(caller.module, :callback, {:callback, expr, caller})
+
+    Kernel.Typespec.ensure_use_behaviour(caller.module)
```

Comment:
The plan is to not do this. The idea is that `__behaviour__(:docs)` will be gone and we will fetch them with `Kernel.Typespec.beam_callbackdocs` (the implementation for callback docs can just call `Code.get_docs` as we do in `__behaviour__(:docs)`). For now, you can also keep the name `behaviour_docs` too. In some later work, I want to remove it to something else and also store typedoc in the bytecode.

Also, we will need tests. I would just recommend getting the current `behaviour_test.exs`, port it to `typespec_test.exs` and ensure we have all functionality, but without using Behaviour at all.